### PR TITLE
add support for aggregate MDS in discovery_read, --agg flag

### PR DESCRIPTION
**Pull request by craigrbarnes**
Jira Ticket: [HP-546](https://ctds-planx.atlassian.net/browse/HP-546)

### New Features
* Add Aggregate metadata service support for discovery read: --agg flag to enable

### CODE REVIEW

**Comment by mcannalte** [comment]
Is this the expected behavior? I'm getting back 587 results compared to the 601 quoted on healdata.org:
```
$ cat $(gen3 --auth ~/creds/heal.json discovery read) | wc -l
501
$ cat $(gen3 --auth ~/creds/heal.json discovery read --agg) | wc -l
587
```

Code looks good to me

File: tests/test_discovery.py [hunk 1]
```diff
@@ -83,6 +81,36 @@ def test_discovery_read(metadata_query_patch, metadata_file_patch, gen3_auth):
...
+
+        outfile.truncate(0)
+        loop.run_until_complete(
+            output_expanded_discovery_metadata(
+                gen3_auth, endpoint="excommons.org", use_agg_mds=True
+            )
+        )
+        outfile.seek(0)
+        csv_rows = list(csv.DictReader(outfile, delimiter="\t"))
+        assert len(csv_rows) == 2
```
Comment (paulineribeyre) [inline tests/test_discovery.py:112]: should we check the row contents as well to make sure the results flattening works as expected?

**Comment by mcannalte** [comment]
Looks like there are some parsing issues (maybe with the guids?) when running this on heal data. I'm going to take a closer look, but here's the issue I'm seeing:
image

File: tests/test_discovery.py (hunk 1 above)
Comment (craigrbarnes) [reply tests/test_discovery.py:112]: Agreed will add additional test.

**Comment by craigrbarnes** [comment]
> Is this the expected behavior? I'm getting back 587 results compared to the 601 quoted on healdata.org:
> 
> ```
> $ cat $(gen3 --auth ~/creds/heal.json discovery read) | wc -l
> 501
> $ cat $(gen3 --auth ~/creds/heal.json discovery read --agg) | wc -l
> 587
> ```
> 
> Code looks good to me
> 
> edit:
> 
> ```
> $ cat $(gen3 --auth ~/creds/heal.json discovery read --agg --limit 1000) | wc -l
> 691
> ```
Yes found the issue and will refactor to deal with it.

File: gen3/tools/metadata/discovery.py [hunk 2]
```diff
@@ -60,17 +89,18 @@ async def output_expanded_discovery_metadata(auth, endpoint=None, limit=500):
...
-                delimiter="\t",
-                quotechar="",
-                quoting=csv.QUOTE_NONE,
-                escapechar="\\",
+                **{**BASE_CSV_PARSER_SETTINGS, "fieldnames": output_columns},
             )
             writer.writeheader()
 
+            a = sorted(os.listdir(metadata_cache_dir))
+            count = 0
```
Comment (mcannalte) [inline gen3/tools/metadata/discovery.py:103]:
```suggestion
```
Looks like these are no longer used, unless they're part of some work still in progress

File: gen3/tools/metadata/discovery.py [hunk 3]
```diff
@@ -79,16 +109,19 @@ async def output_expanded_discovery_metadata(auth, endpoint=None, limit=500):
...
                             **fetched_metadata,
                             **flattened_tags,
-                            "guid": guid,
-                        }.items()
-                    }
+                            "guid": true_guid,
+                        }
+                    )
                     writer.writerow(output_metadata)
+                    count += 1
```
Comment (mcannalte) [inline gen3/tools/metadata/discovery.py:124]:
```suggestion
```

**Review by mcannalte — APPROVED** [review]
Works as expected for me now. I tested with 
```
gen3 --auth ~/creds/dev.json discovery publish $(gen3 --auth ~/creds/heal.json discovery read --agg --limit 1000)
```
And inspect that the metadata in my dev commons looks identical to HEAL.

File: gen3/tools/metadata/discovery.py (hunk 2 above)
Comment (craigrbarnes) [reply gen3/tools/metadata/discovery.py:103]: These were just a means for me to double-check on the counts. I will remove these.

**Pull request merged**